File: eb-mdt/eb-mdt/server/announcements.lua
-- server/announcements.lua

-- Define minimum ranks required to manage announcements per department
local ANNOUNCEMENT_RANKS = {
    ['lspd'] = 11,      -- Sergeant and above
    ['bcso'] = 10,      -- Sergeant and above
    ['sasp'] = 5,      -- Corporal and above
    ['usms'] = 4,      -- Deputy Marshal and above
}

-- Helper function to check if player can manage announcements
local function canManageAnnouncements(player)
    if not player or not player.PlayerData.job.type == 'leo' or not player.PlayerData.job.onduty then
        return false
    end
    
    local jobName = player.PlayerData.job.name:lower()
    local gradeLevel = player.PlayerData.job.grade.level or 0
    local requiredRank = ANNOUNCEMENT_RANKS[jobName]
    
    return requiredRank and gradeLevel >= requiredRank
end

-- Get all active announcements
lib.callback.register('mdt:getAnnouncements', function(source)
    local src = source
    local player = exports.qbx_core:GetPlayer(src)
    
    -- Verify player is LEO and on duty
    if not player or player.PlayerData.job.type ~= 'leo' or not player.PlayerData.job.onduty then
        return nil
    end
    
    -- Get announcements that haven't expired
    local query = [[
        SELECT 
            id,
            title,
            content,
            author,
            author_citizenid,
            department,
            importance,
            created_at,
            expires_at
        FROM mdt_announcements
        WHERE expires_at IS NULL OR expires_at > NOW()
        ORDER BY 
            FIELD(importance, 'critical', 'high', 'medium', 'low'),
            created_at DESC
        LIMIT 50
    ]]
    
    local announcements = exports.oxmysql:fetchSync(query, {})
    
    -- Add permission info for each announcement
    local canManage = canManageAnnouncements(player)
    local playerCitizenId = player.PlayerData.citizenid
    
    if announcements then
        for _, announcement in ipairs(announcements) do
            -- Player can delete if they have permission OR they created the announcement
            announcement.canDelete = canManage or announcement.author_citizenid == playerCitizenId
        end
    end
    
    return {
        announcements = announcements or {},
        canCreate = canManage
    }
end)

-- Create new announcement
lib.callback.register('mdt:createAnnouncement', function(source, data)
    local src = source
    local player = exports.qbx_core:GetPlayer(src)
    
    -- Verify player has permission
    if not canManageAnnouncements(player) then
        return {success = false, message = "Insufficient rank to create announcements"}
    end
    
    -- Validate data
    if not data.title or data.title == '' then
        return {success = false, message = "Title is required"}
    end
    
    if not data.content or data.content == '' then
        return {success = false, message = "Content is required"}
    end
    
    -- Prepare author info
    local authorName = player.PlayerData.charinfo.firstname .. ' ' .. player.PlayerData.charinfo.lastname
    local authorCitizenId = player.PlayerData.citizenid
    local department = player.PlayerData.job.label
    
    -- Calculate expiry if duration is provided (in hours)
    local expiresAt = nil
    if data.duration and tonumber(data.duration) > 0 then
        -- MySQL DATE_ADD equivalent
        expiresAt = os.date('%Y-%m-%d %H:%M:%S', os.time() + (tonumber(data.duration) * 3600))
    end
    
    -- Insert announcement
    local insertQuery = [[
        INSERT INTO mdt_announcements (title, content, author, author_citizenid, department, importance, expires_at)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    ]]
    
    local result = exports.oxmysql:insertSync(insertQuery, {
        data.title,
        data.content,
        authorName,
        authorCitizenId,
        department,
        data.importance or 'medium',
        expiresAt
    })
    
    if result then
        -- Log the action
        TriggerEvent('mdt:server:log', {
            action = 'Announcement Created',
            officer = authorName,
            details = 'Title: ' .. data.title .. ' | Importance: ' .. (data.importance or 'medium')
        })
        
        -- Notify all online officers
        local players = exports.qbx_core:GetQBPlayers()
        for _, targetPlayer in pairs(players) do
            if targetPlayer and targetPlayer.PlayerData and targetPlayer.PlayerData.job.type == 'leo' and targetPlayer.PlayerData.job.onduty then
                TriggerClientEvent('mdt:announcementCreated', targetPlayer.PlayerData.source, {
                    title = data.title,
                    importance = data.importance or 'medium',
                    author = authorName
                })
            end
        end
        
        return {success = true, id = result}
    end
    
    return {success = false, message = "Failed to create announcement"}
end)

-- Delete announcement
lib.callback.register('mdt:deleteAnnouncement', function(source, announcementId)
    local src = source
    local player = exports.qbx_core:GetPlayer(src)
    
    if not player or player.PlayerData.job.type ~= 'leo' or not player.PlayerData.job.onduty then
        return {success = false, message = "Not authorized"}
    end
    
    -- Get announcement details first
    local checkQuery = [[
        SELECT author_citizenid, title
        FROM mdt_announcements
        WHERE id = ?
    ]]
    
    local announcement = exports.oxmysql:fetchSync(checkQuery, {announcementId})
    
    if not announcement or not announcement[1] then
        return {success = false, message = "Announcement not found"}
    end
    
    local canManage = canManageAnnouncements(player)
    local isAuthor = announcement[1].author_citizenid == player.PlayerData.citizenid
    
    -- Check if player can delete (has permission OR is the author)
    if not canManage and not isAuthor then
        return {success = false, message = "You can only delete your own announcements"}
    end
    
    -- Delete the announcement
    local deleteQuery = [[
        DELETE FROM mdt_announcements
        WHERE id = ?
    ]]
    
    local result = exports.oxmysql:executeSync(deleteQuery, {announcementId})
    
    if result and result.affectedRows and result.affectedRows > 0 then
        local officerName = player.PlayerData.charinfo.firstname .. ' ' .. player.PlayerData.charinfo.lastname
        
        -- Log the action
        TriggerEvent('mdt:server:log', {
            action = 'Announcement Deleted',
            officer = officerName,
            details = 'Title: ' .. announcement[1].title
        })
        
        -- Notify all online officers to refresh
        TriggerClientEvent('mdt:refreshAnnouncements', -1)
        
        return {success = true}
    end
    
    return {success = false, message = "Failed to delete announcement"}
end)

-- Update announcement (for future implementation)
lib.callback.register('mdt:updateAnnouncement', function(source, data)
    local src = source
    local player = exports.qbx_core:GetPlayer(src)
    
    if not player or not canManageAnnouncements(player) then
        return {success = false, message = "Not authorized"}
    end
    
    -- Get announcement to check ownership
    local checkQuery = [[
        SELECT author_citizenid
        FROM mdt_announcements
        WHERE id = ?
    ]]
    
    local announcement = exports.oxmysql:fetchSync(checkQuery, {data.id})
    
    if not announcement or not announcement[1] then
        return {success = false, message = "Announcement not found"}
    end
    
    -- Only author or high-ranking officers can edit
    local isAuthor = announcement[1].author_citizenid == player.PlayerData.citizenid
    if not isAuthor and not canManageAnnouncements(player) then
        return {success = false, message = "Not authorized to edit this announcement"}
    end
    
    local updateQuery = [[
        UPDATE mdt_announcements
        SET title = ?, content = ?, importance = ?
        WHERE id = ?
    ]]
    
    local result = exports.oxmysql:executeSync(updateQuery, {
        data.title,
        data.content,
        data.importance,
        data.id
    })
    
    if result and result.affectedRows and result.affectedRows > 0 then
        TriggerClientEvent('mdt:refreshAnnouncements', -1)
        return {success = true}
    end
    
    return {success = false, message = "Failed to update announcement"}
end)

-- Clean up expired announcements (run periodically)
CreateThread(function()
    while true do
        Wait(300000) -- Check every 5 minutes
        
        local deleteQuery = [[
            DELETE FROM mdt_announcements
            WHERE expires_at IS NOT NULL AND expires_at <= NOW()
        ]]
        
        local result = exports.oxmysql:executeSync(deleteQuery, {})
        
        if result and result.affectedRows and result.affectedRows > 0 then
            print(('[MDT] Cleaned up %d expired announcements'):format(result))
            -- Notify clients to refresh if any were deleted
            TriggerClientEvent('mdt:refreshAnnouncements', -1)
        end
    end
end)

File: src/server/announcements.ts
// Announcements server module
// Uses SQLite for mdt_announcements table
import Config from '@common/config';
import { DebugLog } from '@common/index';
import { onClientCallback } from '@communityox/ox_lib/server';
import { SQLite, extractParam } from './db';
import type { QBXPlayer, Announcement, AnnouncementsResponse, ApiResponse } from '@common/types';

// ==========================================
// Helper Functions
// ==========================================

function isPlayerAuthorized(source: number): QBXPlayer | null {
  const player = exports.qbx_core.GetPlayer(source) as QBXPlayer | null;
  if (!player || player.PlayerData.job.type !== 'leo' || !player.PlayerData.job.onduty) {
    return null;
  }
  return player;
}

function canManageAnnouncements(player: QBXPlayer): boolean {
  const jobName = player.PlayerData.job.name.toLowerCase();
  const gradeLevel = player.PlayerData.job.grade.level || 0;
  const requiredRank = Config.RankRequirements.AnnouncementManagement[jobName as keyof typeof Config.RankRequirements.AnnouncementManagement];

  return requiredRank !== undefined && gradeLevel >= requiredRank;
}

// ==========================================
// Get Announcements
// ==========================================

onClientCallback('mdt:getAnnouncements', (source: number): AnnouncementsResponse | null => {
  const player = isPlayerAuthorized(source);
  if (!player) return null;

  try {
    // SQLite uses different syntax for date comparison and ordering
    const announcements = SQLite.query<Announcement>(
      `SELECT
         id, title, content, author, author_citizenid, department,
         importance, created_at, expires_at
       FROM mdt_announcements
       WHERE expires_at IS NULL OR expires_at > datetime('now')
       ORDER BY
         CASE importance
           WHEN 'critical' THEN 1
           WHEN 'high' THEN 2
           WHEN 'medium' THEN 3
           WHEN 'low' THEN 4
           ELSE 5
         END,
         created_at DESC
       LIMIT ?`,
      [Config.QueryLimits.Announcements]
    );

    const canManage = canManageAnnouncements(player);
    const playerCitizenId = player.PlayerData.citizenid;

    // Add canDelete property
    const announcementsWithPermissions = (announcements || []).map(announcement => ({
      ...announcement,
      canDelete: canManage || announcement.author_citizenid === playerCitizenId,
    }));

    return {
      announcements: announcementsWithPermissions,
      canCreate: canManage,
    };
  } catch (error) {
    console.error('[MDT] Error fetching announcements:', error);
    return { announcements: [], canCreate: false };
  }
});

// ==========================================
// Create Announcement
// ==========================================

onClientCallback('mdt:createAnnouncement', (source: number, data: {
  title: string;
  content: string;
  importance?: string;
  duration?: number;
}): ApiResponse => {
  const player = isPlayerAuthorized(source);
  if (!player) return { success: false, message: 'Not authorized' };

  if (!canManageAnnouncements(player)) {
    return { success: false, message: 'Insufficient rank to create announcements' };
  }

  if (!data.title || data.title === '') {
    return { success: false, message: 'Title is required' };
  }

  if (!data.content || data.content === '') {
    return { success: false, message: 'Content is required' };
  }

  const authorName = `${player.PlayerData.charinfo.firstname} ${player.PlayerData.charinfo.lastname}`;
  const authorCitizenId = player.PlayerData.citizenid;
  const department = player.PlayerData.job.label;

  // Calculate expiry if duration provided (in hours)
  let expiresAt: string | null = null;
  if (data.duration && Number(data.duration) > 0) {
    const expiryTime = new Date(Date.now() + Number(data.duration) * 3600 * 1000);
    expiresAt = expiryTime.toISOString().slice(0, 19).replace('T', ' ');
  }

  try {
    const result = SQLite.insert(
      `INSERT INTO mdt_announcements (title, content, author, author_citizenid, department, importance, expires_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [
        data.title,
        data.content,
        authorName,
        authorCitizenId,
        department,
        data.importance || Config.Defaults.AnnouncementImportance,
        expiresAt,
      ]
    );

    if (result) {
      emit('mdt:server:log', {
        action: 'Announcement Created',
        officer: authorName,
        details: `Title: ${data.title} | Importance: ${data.importance || 'medium'}`,
      });

      // Notify all online officers
      const players = exports.qbx_core.GetQBPlayers() as Record<number, QBXPlayer>;
      for (const [, targetPlayer] of Object.entries(players)) {
        if (targetPlayer?.PlayerData?.job?.type === 'leo' && targetPlayer.PlayerData.job.onduty) {
          emitNet('mdt:announcementCreated', targetPlayer.PlayerData.source, {
            title: data.title,
            importance: data.importance || 'medium',
            author: authorName,
          });
        }
      }

      return { success: true, data: { id: result } };
    }

    return { success: false, message: 'Failed to create announcement' };
  } catch (error) {
    console.error('[MDT] Error creating announcement:', error);
    return { success: false, message: 'Database error' };
  }
});

// ==========================================
// Delete Announcement
// ==========================================

onClientCallback('mdt:deleteAnnouncement', (source: number, data: unknown): ApiResponse => {
  const player = isPlayerAuthorized(source);
  if (!player) return { success: false, message: 'Not authorized' };

  const announcementId = typeof data === 'number' ? data : extractParam<number>(data, 'announcementId') || 0;
  if (!announcementId) return { success: false, message: 'Invalid announcement ID' };

  // Get announcement details
  const announcement = SQLite.single<{ author_citizenid: string; title: string }>(
    'SELECT author_citizenid, title FROM mdt_announcements WHERE id = ?',
    [announcementId]
  );

  if (!announcement) {
    return { success: false, message: 'Announcement not found' };
  }

  const canManage = canManageAnnouncements(player);
  const isAuthor = announcement.author_citizenid === player.PlayerData.citizenid;

  if (!canManage && !isAuthor) {
    return { success: false, message: 'You can only delete your own announcements' };
  }

  try {
    SQLite.update('DELETE FROM mdt_announcements WHERE id = ?', [announcementId]);

    const officerName = `${player.PlayerData.charinfo.firstname} ${player.PlayerData.charinfo.lastname}`;

    emit('mdt:server:log', {
      action: 'Announcement Deleted',
      officer: officerName,
      details: `Title: ${announcement.title}`,
    });

    // Notify clients to refresh
    emitNet('mdt:refreshAnnouncements', -1);

    return { success: true };
  } catch (error) {
    console.error('[MDT] Error deleting announcement:', error);
    return { success: false, message: 'Database error' };
  }
});

// ==========================================
// Update Announcement
// ==========================================

onClientCallback('mdt:updateAnnouncement', (source: number, data: {
  id: number;
  title: string;
  content: string;
  importance: string;
}): ApiResponse => {
  const player = isPlayerAuthorized(source);
  if (!player) return { success: false, message: 'Not authorized' };

  const announcement = SQLite.single<{ author_citizenid: string }>(
    'SELECT author_citizenid FROM mdt_announcements WHERE id = ?',
    [data.id]
  );

  if (!announcement) {
    return { success: false, message: 'Announcement not found' };
  }

  const isAuthor = announcement.author_citizenid === player.PlayerData.citizenid;
  if (!isAuthor && !canManageAnnouncements(player)) {
    return { success: false, message: 'Not authorized to edit this announcement' };
  }

  try {
    SQLite.update(
      'UPDATE mdt_announcements SET title = ?, content = ?, importance = ? WHERE id = ?',
      [data.title, data.content, data.importance, data.id]
    );

    emitNet('mdt:refreshAnnouncements', -1);

    return { success: true };
  } catch (error) {
    console.error('[MDT] Error updating announcement:', error);
    return { success: false, message: 'Database error' };
  }
});

// ==========================================
// Cleanup Expired Announcements
// ==========================================

// Run every 5 minutes
setInterval(() => {
  try {
    const result = SQLite.update(
      "DELETE FROM mdt_announcements WHERE expires_at IS NOT NULL AND expires_at <= datetime('now')",
      []
    );

    if (result && result > 0) {
      console.log(`[MDT] Cleaned up ${result} expired announcements`);
      emitNet('mdt:refreshAnnouncements', -1);
    }
  } catch (error) {
    console.error('[MDT] Error cleaning up expired announcements:', error);
  }
}, Config.RefreshIntervals.ExpiredAnnouncements);
